Changes related to Linux-based build.

ValMenn opened the pull request:
1. Moved most of the NuGet-related tasks into src/.nuget/packageLoad.targets
2. Since NuGet.exe cannot process solutions on Linux created an option
  to use dnu for project.json processing
3. Copy CoreRun to the output directory in Linux
4. Fixed a number of problems with test processes building NetCore configuration
5. Added a few generic functions to replace Window-specific ones
6. Updated some package versions

It still does not quite builds on Linux because linux runtime cannot be added project.json

AndyGerlicher commented on dir.props:
```diff
@@ -145,17 +140,22 @@
   </PropertyGroup>
   <PropertyGroup Condition="'$(NetCoreBuild)' == 'true'">
     <TargetFrameworkVersion>v5.0</TargetFrameworkVersion>
-    <PlatformTarget>x86</PlatformTarget>
```
Was there a particular reason to drop specifying x86?

ValMenn replied:
I moved it down, it's not gone

weshaggard replied:
Actually is there any reason that we cannot use AnyCPU for these task libraries? We are having any issue right now for example where we are trying to build and AnyCPU tasks library that we can use for both windows and linux but we cannot because of this split.

AndyGerlicher commented on dir.props:
```diff
@@ -145,17 +140,22 @@
   <PropertyGroup Condition="'$(NetCoreBuild)' == 'true'">
     <TargetFrameworkVersion>v5.0</TargetFrameworkVersion>
-    <PlatformTarget>x86</PlatformTarget>
+    <TargetFrameworkIdentifier>DNXCore</TargetFrameworkIdentifier>
```
Is this right? For Visual Studio?

ValMenn replied:
Why is it wrong for CoreCLR build?

AndyGerlicher commented on src/.nuget/NuGet.Config:
```diff
@@ -6,6 +6,8 @@
     <add key="myget.org dotnet-corefx" value="https://www.myget.org/F/dotnet-corefx/" />
+    <add key="myget.org nugetbuild" value="https://www.myget.org/F/nugetbuild/api/v3/index.json" />
+    <add jey="nuget.org v2" value="https://www.nuget.org/api/v2/index.json" />
     <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
```
Is this file still used for NuGet v3?

ValMenn replied:
I specify it on the command line. If you know why I should not, please let me know.

AndyGerlicher commented:
Looks like the CI is failing: EXEC : warning : Unable to find version '1.0.25-prerelease-00098' of package 'Microsoft.DotNet.BuildTools'

AndyGerlicher commented on dir.props:
```diff
@@ -25,9 +25,11 @@
     <BuildToolsVersion>1.0.25-prerelease-00080</BuildToolsVersion>
```
This looks like it should be updated (or removed if it's not used).

ValMenn commented:
I did notice that build failed. I think the source is missing, but it works both on my Windows and Linux systems (it might have found the source in a global config, or something). I'll see how I can repro the failure on one of them.

AndyGerlicher commented:
It definitely fails for me (Windows) after `git clean -xdf`.

rainersigwald commented on Samples/dir.targets:
```diff
@@ -5,10 +5,24 @@
-  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\Portable\$(TargetFrameworkVersion)\Microsoft.Portable.CSharp.targets" Condition="'$(NetCoreBuild)' == 'true'"/>
+  <Import Project="$(NugetIntegrationDir)\Microsoft.NuGet.targets" Condition="'$(OsEnvironment)'=='Unix'" />
 
+  <Target Name="ResolveNuGetPackageAssets" Condition="exists('$(ProjectLockFile)')">
```
Doesn't this come with Microsoft.NuGet.targets?  Are you overriding it?

ValMenn replied:
Yes

dsplaisted replied:
Can you change this so it only overrides when building on Unix?  The task DLL in the NuGet.MsBuild.Integration package is test or delay signed, so I get a strong name validation error when I try to build using it on Windows.

rainersigwald commented on build.proj:
```diff
@@ -18,8 +18,9 @@
           BeforeTargets="Build;BuildAllProjects"
-          DependsOnTargets="_RestoreBuildTools">
+          DependsOnTargets="_RestoreBuildTools"
+          Condition="'$(SolutionFile)'!='' and Exists('$(SolutionFile)')">
```
This seems a bit too tolerant.  What valid situation do we have where this won't exist but everything's ok?

ValMenn replied:
If the SolutionFile is not set, nuget.exe won't process the solution file. In that case the targets will process it using dnu.

rainersigwald commented on dir.targets:
```diff
@@ -1,154 +1,40 @@
-
-  <Target Name="_CopyCompilers" AfterTargets="_RestoreBuildTools" Condition="'$(UseRoslynCompilers)' == 'true'">
+  <Target Name="_CopyCompilers"
+          BeforeTargets="_CopyDataSetPermissions"
```
What does this do?  Can you structure it as a DependsOnTargets in the other target instead if it's still necessary?

ValMenn replied:
You're right, fixed.

rainersigwald commented on src/.nuget/NuGet.Config:
```diff
@@ -3,14 +3,14 @@
   <packageRestore>
     <add key="enabled" value="True" />
   </packageRestore>
+  <packageSources>
```
Do we actually need these listed both here and in the root?

ValMenn replied:
I didn't even notice there was a NuGet.Config in the root. I don't know why it's there.

rainersigwald commented on src/.nuget/packages.config:
```diff
@@ -1,7 +1,8 @@
 <?xml version="1.0" encoding="utf-8"?>
 <packages>
```
Maybe add a comment here for what goes in the .xml list and what goes in the .json?  I don't understand :confused:

ValMenn replied:
This is for Linux and I used it to specify runtime (I need different for the different systems, it seems). Still not sure how to deal with this.

rainersigwald commented on src/.nuget/packageLoad.targets:
```diff
@@ -0,0 +1,150 @@
+          InstallationDependency - list of files that, when changed, need to cause reinstall
+          BuildToolsSemaphore    - Created when build tools are restored
+        1. Downloads NuGet.exe
+        2. Handle packages.config and download packages it specifies
```
Can you be more concrete than "handle"?  This note *almost* tells me what I need to know but I still have to read the targets.

rainersigwald commented on src/.nuget/packageLoad.targets:
```diff
@@ -0,0 +1,150 @@
+
+  <!-- These are properties for the dnu -->
+  <PropertyGroup>
+    <PackagesCache Condition="'$(OsEnvironment)'=='Windows_NT'">$([System.IO.Path]::Combine($(USERPROFILE), '.nuget', 'packages'))</PackagesCache>
```
Under what circumstances do we need dnu on windows?

ValMenn replied:
If I add Linux runtime to project.json files, nuget.exe on the solution fails. Running dnu succeeds. Since I did not add Linux runtime in the PR, this is not enabled and we currently don't use dnu on Windows, but it may (potentially) be enabled in the next PR unless I find how to use nuget.exe.

dsplaisted commented on build.proj:
```diff
@@ -5,7 +5,7 @@
   </PropertyGroup>
   
-  <Import Project="dir.props" />
+  <Import Project="src\dir.props" />
```
Is there any reason not to move the PropertyGroup you added to `src\dir.props` into the root `dir.props` file?  Then you could continue importing `dir.props` from the root here, which seems a bit less convoluted and more in line with the pattern of how `dir.props` is used elsewher.

ValMenn replied:
Well, the idea is that when we are building projects from src, we need to process the .sln file, while projects in Samples have no solution. So these properties do belong in the src\dir.props.

dsplaisted replied:
The samples are included in the .sln under src, even though they are in a different directory.  So they have the same solution as the rest of the projects.

I do think we should consider moving the samples folder under `src` (as a separate task).  Right now we're duplicating things between the dir.targets under `src/` and `Samples/`.

ValMenn replied:
I'll look into it for the next batch of changes

rainersigwald commented on src/Shared/NativeMethodsShared.cs:
```diff
@@ -918,25 +918,56 @@ internal static void KillTree(int processIdToKill)
+                        line = r.ReadLine();
+                    }
+                }
+                catch // Ignore errors since the process may have terminated
```
Is it possible to be more specific in the catch?  Swallowing all exceptions gives me the heebie-jeebies.

ValMenn replied:
I don't think it's possible at this time. I have no idea what kind of exception I may get. I may clarify this after some experimentation.

rainersigwald commented on src/Shared/NativeMethodsShared.cs:
```diff
@@ -918,25 +918,56 @@ internal static void KillTree(int processIdToKill)
+                {
+                    // One of the fields is the process name. It may contain any characters, but since it's in
+                    // parenthesis, we can finds its end by looking for the last parenthesis. Afther that, the
+                    // comes a space, then the second fields seeparated by a space is the parent id.
```
typo: separated

rainersigwald commented on src/XMakeBuildEngine/UnitTests/BackEnd/LoggingService_Tests.cs:
```diff
@@ -381,7 +381,7 @@ public void RegisterDistributedLoggerServiceShutdown()
                 LoggerDescription description = CreateLoggerDescription(className, Assembly.GetAssembly(typeof(ProjectCollection)).FullName, true);
 #else
-            LoggerDescription description = CreateLoggerDescription(className, typeof(ProjectCollection).GetTypeInfo().Assembly.FullName, true);
+                LoggerDescription description = CreateLoggerDescription(className, typeof(ProjectCollection).GetTypeInfo().Assembly.FullName, true);
```
Low-pri but I'll ask anyway:  can you move the whitespace-only changes into their own commit so future history spelunking is easier?

rainersigwald commented on src/XMakeTasks/NativeMethods.cs:
```diff
@@ -967,6 +967,22 @@ internal static bool MoveFileEx(string existingFileName, string newFileName, Mov
+
+        internal static bool AllDrivesMapped()
+        {
+            const uint AllDriveMask = 0x0cffffff;
```
Why use this approach rather than the other `Length > 26` approach?

rainersigwald commented on src/XMakeTasks/NativeMethods.cs:
```diff
@@ -967,6 +967,22 @@ internal static bool MoveFileEx(string existingFileName, string newFileName, Mov
+        internal static bool AllDrivesMapped()
+        {
+            const uint AllDriveMask = 0x0cffffff;
+            if (NativeMethodsShared.IsWindows)
```
I think I prefer something like

```C#
if (!NativeMethodsShared.IsWindows)
{
    // No limit to drive mount points
    return false;
}
```

rainersigwald commented on src/XMakeTasks/Exec.cs:
```diff
@@ -466,7 +466,7 @@ protected override bool ValidateParameters()
             // if the working directory is a UNC path, and all drive letters are mapped, bail out, because the pushd command
             // will not be able to auto-map to the UNC path
-            if (workingDirectoryIsUNC && (DriveInfo.GetDrives().Length == 26))
+            if (workingDirectoryIsUNC && NativeMethods.AllDrivesMapped())
```
does `workingDirectoryIsUNC` even have meaning on Unix?  Should this be resolved by making that always false?

dsplaisted commented on Samples/dir.targets:
```diff
@@ -5,10 +5,24 @@
   </PropertyGroup>
   
   <Import Project="..\dir.targets" />
+  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets"/>
```
When using Microsoft.CSharp.targets instead of Microsoft.Portable.CSharp.targets, I get the following warning:

> C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(1097,5): warning MSB3644: The reference assemblies for framework "DNXCore,Version=v5.0" were not found. To resolve this, install the SDK or Targeting Pack for this framework version or retarget your application to a version of the framework for which you have the SDK or Targeting Pack installed. Note that assemblies will be resolved from the Global Assembly Cache (GAC) and will be used in place of reference assemblies. Therefore your assembly may not be correctly targeted for the framework you intend. [C:\git\msbuild\src\Framework\Microsoft.Build.Framework.csproj]

I think we'll need a NuGet package that provides the portable targets for us.  For now, can you make it so that when building on Windows and `NetCoreBuild` is true, that it will import `Microsoft.Portable.CSharp.targets` instead of `Microsoft.CSharp.targets`?

rainersigwald commented on src/dir.targets:
```diff
@@ -51,6 +45,9 @@
-  
+
+  <PropertyGroup Condition="'$(NetCoreBuild)' == 'true'">
+    <NuGetTargetMoniker>DNXCore,Version=v5.0</NuGetTargetMoniker>
```
Why does this need to be set after the use in ResolveAssets? (comment?)

rainersigwald commented on src/dir.targets:
```diff
@@ -5,12 +5,7 @@
   </PropertyGroup>
   
-  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" Condition="'$(NetCoreBuild)' != 'true'"/>
-  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\Portable\$(TargetFrameworkVersion)\Microsoft.Portable.CSharp.targets" Condition="'$(NetCoreBuild)' == 'true'"/>
```
How does the portable/not thing work now?

ValMenn commented:
Broke the changes up into two commits per Rainer's suggestion. Made some changes based on recommendations in the comments. Specifically, made some changes Windows-specific (such as task override and used Microsoft.Portable.CSharp.targets on Windows.

dsplaisted commented:
It looks like you've removed the NuGet.config in the repository root.  I think we should leave that one there, and remove the one in `src\.nuget`.  The reason for this is because NuGet will automatically walk up the parent directories of the project directory looking for NuGet.config files.  So for example if our NuGet.config is at the repo root, then the NuGet UI and package restore in Visual Studio will work with the feeds we've set up there.

dsplaisted commented:
I tried building your changes on Windows, and it looks like it's no longer downloading NuGet.exe if it's not found in the packages folder.

dsplaisted commented on src/.nuget/packageLoad.targets:
```diff
@@ -0,0 +1,156 @@
+  <!-- Restore NuGet.exe and use it to process Nuget.config. -->
+  <Target Name="_RestoreBuildToolsNugetConfig"
+          Inputs="$(GlobalPropertiesFile);$(NuGetPackagesFile)"
+          Outputs="$(BuildToolsSemaphore);$(BuildToolsSemaphore)"
```
I think this should be

    Outputs="$(BuildToolsSemaphore);$(NuGetToolPath)"

Instead of:

    Outputs="$(BuildToolsSemaphore);$(BuildToolsSemaphore)"

Otherwise the target is skipped because it thinks everything is up to date and NuGet.exe isn't downloaded.

dsplaisted commented:
LGTM, modulo my comments about leaving NuGet.Config in the repo root and fixing the target which is supposed to download NuGet.exe.

AndyGerlicher commented:
Things still don't seem right here. Few things I've noticed running locally (on the 83883ee commit):
* The build seems significantly slower than xplat tip (30% slower on command line, and noticeably in VS). 
* The CI build is failing about 1000 more tests than previous (but that could be just older branch). 
* Error in VS building (that don't appear in tip):
```
Couldn't find the required information in the lock file. Make sure you have  NETFramework,Version=v4.5.1/win7-x86 mentioned in your targets.
```

* After build the following lock files are modified:
```
        modified:   Samples/PortableTask/project.lock.json
        modified:   src/Utilities/UnitTests/project.lock.json
        modified:   src/Utilities/project.lock.json
        modified:   src/XMakeBuildEngine/UnitTests/project.lock.json
        modified:   src/XMakeBuildEngine/project.lock.json
        modified:   src/XMakeCommandLine/UnitTests/project.lock.json
        modified:   src/XMakeCommandLine/project.lock.json
        modified:   src/XMakeTasks/UnitTests/project.lock.json
        modified:   src/XMakeTasks/project.lock.json
        modified:   src/nuget/Microsoft.Build.nuspec
```

At this point would it make sense to break things up a little so we can isolate them a bit more? The native code methods and of the minor changes look good so it would be great to get those in. I'd like to understand more the differences and if they're good or not before taking the others (*sigh* NuGet complexity...).

ValMenn commented:
* Moved *NuGet.Config* to the root
* Changed all imports in *dir.targets* so that the import only happens when the target file exists
* A few more changes to the project files

One must still do command line build before building in Visual Studio, but the build succeed.

AndyGerlicher commented:
@ValMenn This is merged, and I get fail in build:
```
D:\src\msbuild.fork\dir.targets(176,5): error MSB3073: The command "D:\src\msbuild.fork\packages\\xunit.runner.console.2.1.0-rc1-build3168\tools\xunit.console.x86.exe D:\src\msbuild.fork\bin\Windows_NT\Debug\Microsoft.Build.Tasks.UnitTests.dll -noshadow -parallel none -xml D:\src\msbuild.fork\bin\Windows_NT\Debug\Microsoft.Build.Tasks.UnitTests.dll_TestResults.xml -html D:\src\msbuild.fork\bin\Windows_NT\Debug\Microsoft.Build.Tasks.UnitTests.dll_TestResults.html > D:\src\msbuild.fork\bin\Windows_NT\Debug\Microsoft.Build.Tasks.UnitTests.dll_stdout.txt" exited with code 1.
```

Looking in that file:
```
xUnit.net Console Runner (32-bit .NET 4.0.30319.42000)
System.IO.FileLoadException: Could not load file or assembly 'xunit.core, Version=2.1.0.3168, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
```

Did this pass CI? Or work on your machine?

rainersigwald commented:
@AndyGerlicher looks like we broke tests for full framework in xplat at some point; I just saw this on my machine.  Rolling build for xplat seems as happy as ever: http://dotnet-ci.cloudapp.net/job/microsoft_msbuild_xplat/60/ (it only tests netcore).

I'll see what I can figure out.

rainersigwald commented:
@AndyGerlicher I'm stumped.  I can no longer repro, including on a clean VM.  Can you?  If so, can you try https://github.com/rainersigwald/msbuild/tree/revive-fullframework-tests and see if that helps?
